### PR TITLE
Support pinning of IPythonBuffer

### LINKED ISSUE #1297: Support pinning of IPythonBuffer
Currently the `IPythonBuffer` interface only provides access to the buffer via a Span<byte>. Adding `IPythonBuffer.AsMemory<byte>()` would enable some scenarios such as [`ctypes._CData.from_buffer`](https://docs.python.org/3/library/ctypes.html#ctypes._CData.from_buffer) which requires(?) pinning of the memory for use in unmanaged code.

@BCSharp since you implemented `IPythonBuffer`, any thoughts or concerns?

### PR DESCRIPTION
**Pull request by BCSharp**
Resolves #1297 though more can be done. Currently the lifetime of `MemoryHolder` is not managed efficiently when used from C#. It was like that before this change and it does not affect Python code, which relies on the finalizers and the GC, but for C# calls, it can be done more efficiently using `Dispose()`. I think there are more improvements/modernizations/cleanups possible here, and likely in the whole `ctypes` module, but that is for another PR (probably several).

I have locally other changes in `ctypes` that would address the remaining part from #1404 and I used them for testing, not sure should they go in the separate PR (since they address a separate issue) or here (since they validate the code changes).

### CODE REVIEW

**Comment by slozier** [comment]
Looks like a pretty clean solution. I like it.

> I have locally other changes in `ctypes` that would address the remaining part from #1404 and I used them for testing, not sure should they go in the separate PR (since they address a separate issue) or here (since they validate the code changes).

I'm fine either way if the changes are small and doesn't make it hard to review this PR.

File: Src/IronPython/Runtime/MemoryView.cs [hunk 1]
```diff
@@ -930,7 +931,7 @@ public sealed class MemoryView : ICodeFormattable, IWeakReferenceable, IBufferPr
                 throw PythonOps.BufferError("memoryview: underlying buffer is not writable");
 
             if (flags.HasFlag(BufferFlags.CContiguous) && !_isCContig)
-                throw PythonOps.BufferError("memoryview: underlying buffer is not C-contiguous");
+                throw PythonOps.BufferError("memoryview: underlying buffer is not C contiguous");
```
Comment (slozier) [inline Src/IronPython/Runtime/MemoryView.cs:934]: I thought Python used "C-contiguous" with a dash?
```py
struct.unpack("H", memoryview(b"AA")[::-1])
```
> BufferError: memoryview: underlying buffer is not C-contiguous

**Review by slozier — COMMENTED** [review]
Generally looks good to me (other than the change in memoryview error messages).

File: Src/IronPython/Runtime/MemoryView.cs (hunk 1 above)
Comment (BCSharp) [reply Src/IronPython/Runtime/MemoryView.cs:934]: `memoryview` does, but `ctypes` doesn't.
```python
c_short.from_buffer(memoryview(bytearray(b"AA"))[::-1])
```
> TypeError: underlying buffer is not C contiguous

Unfortunately, tests for `ctypes` check for this form. Will see if I can get it both ways.

Comment (slozier) [reply Src/IronPython/Runtime/MemoryView.cs:934]: Took a peek at the CPython ctypes implementation and it seems like `from_buffer` is the odd one:
```
ctypes.c_int.from_buffer(memoryview(bytearray(b"aaaa"))[::-1])
```
> TypeError: underlying buffer is not C contiguous

but `from_buffer_copy` gives the `memoryview` error:
```
ctypes.c_int.from_buffer_copy(memoryview(bytearray(b"aaaa"))[::-1])
```
> BufferError: memoryview: underlying buffer is not C-contiguous

**Comment by slozier** [comment]
Thanks!

**Pull request merged**